src/transport.c

```diff
@@ -1658,46 +1639,69 @@ mama_status zmqBridgeMamaTransportImpl_stopMonitor(zmqTransportBridge* impl)
    return MAMA_STATUS_OK;
 }
 
-typedef struct __attribute__ ((packed)) zmq_monitor_frame1 {
-   uint16_t    event;
-   uint32_t    value;
-} zmq_monitor_frame1;
-
-int zmqBridgeMamaTransportImpl_monitorEvent(void *socket, const char* socketName)
+uint64_t zmqBridgeMamaTransportImpl_monitorEvent_v2(void *socket, const char* socketName)
 {
-   // First frame in message contains event number and value
-   zmq_msg_t msg;
-   zmq_msg_init (&msg);
-   if (zmq_msg_recv (&msg, socket, 0) == -1)
-      return -1; // Interrupted, presumably
-
-   zmq_monitor_frame1* pFrame1 = (zmq_monitor_frame1*) zmq_msg_data (&msg);
-   int event = pFrame1->event;
-   int value = pFrame1->value;
-   const char* eventName = get_zmqEventName(event);
-
-   // Second frame in message contains event address
-   zmq_msg_init (&msg);
-   if (zmq_msg_recv (&msg, socket, 0) == -1)
-      return -1; // Interrupted, presumably
-
-   uint8_t* data = (uint8_t *) zmq_msg_data (&msg);
-   size_t size = zmq_msg_size(&msg);
-   char endpoint[ZMQ_MAX_ENDPOINT_LENGTH +1];
-   memset(endpoint, '\0', sizeof(endpoint));
-   memcpy(endpoint, data, size);
+    //  First frame in message contains event number
+    zmq_msg_t msg;
+    zmq_msg_init (&msg);
+    if (zmq_msg_recv (&msg, socket, 0) == -1)
+        return -1;              //  Interrupted, presumably
+    //assert (zmq_msg_more (&msg));
+
+    uint64_t event;
+    memcpy (&event, zmq_msg_data (&msg), sizeof (event));
+    zmq_msg_close (&msg);
+
+    //  Second frame in message contains the number of values
+    zmq_msg_init (&msg);
+    if (zmq_msg_recv (&msg, socket, 0) == -1)
+        return -1;              //  Interrupted, presumably
+    //assert (zmq_msg_more (&msg));
+
+    uint64_t value_count;
+    memcpy (&value_count, zmq_msg_data (&msg), sizeof (value_count));
+    zmq_msg_close (&msg);
+
+    uint64_t value;
+    for (uint64_t i = 0; i < value_count; ++i) {
+        //  Subsequent frames in message contain event values
+        zmq_msg_init (&msg);
+        if (zmq_msg_recv (&msg, socket, 0) == -1)
+            return -1;              //  Interrupted, presumably
+        //assert (zmq_msg_more (&msg));
+
+        memcpy (&value, zmq_msg_data (&msg), sizeof (value));
```

Comment:
I'm assuming this loop intends to only take the last event?

Reply:
Yes -- and iirc that is the also the first event.  The sample code (not online, but in the file zmq_socket_monitor_versioned.txt) shows an array of events, but in my tests there was only one.

I also commented out assert's, since we don't ever want to do that.